crates/noirc_frontend/src/hir/def_collector/dc_crate.rs

```diff
@@ -601,6 +660,55 @@ fn resolve_impls(
     file_method_ids
 }
 
+fn resolve_trait_impls(
+    context: &mut Context,
+    traits: TraitImplMap,
+    crate_id: CrateId,
+    errors: &mut Vec<FileDiagnostic>,
+) -> Vec<(FileId, FuncId)> {
+    let interner = &mut context.def_interner;
+    let mut methods = Vec::<(FileId, FuncId)>::new();
+
+    for ((unresolved_type, _, trait_id), trait_impl) in traits {
+        let local_mod_id = trait_impl.module_id;
+        let module_id = ModuleId { krate: crate_id, local_id: local_mod_id };
+        let path_resolver = StandardPathResolver::new(module_id);
+
+        let self_type = {
+            let mut resolver =
+                Resolver::new(interner, &path_resolver, &context.def_maps, trait_impl.file_id);
+            resolver.resolve_type(unresolved_type.clone())
+        };
+
+        let mut impl_methods = resolve_function_set(
+            interner,
+            crate_id,
+            &context.def_maps,
+            trait_impl.methods.clone(),
```

Comment:
Can we remove the `#[derive(Clone)]` for unresolved functions and unresolved traits if we remove the clone here? I think `resolve_function_set` does not actually need to take its arguments by value.

Reply:
There's a few structures which we clone very liberally, which I believe we can avoid cloning at all - mainly NoirFunction (which is 300 bytes + has a bunch of Vecs) as well as the AST nodes for Trait/TraitImpl.

Doing this would require changing quite a few method signatures, so it's probably a PR of its own, and we've already got a few PRs planned, so if you want I can try to do it after we get the following merged:

1. Trait impl methods signature checks PR: https://github.com/noir-lang/noir/pull/2652
2. Trait Bounds DC+Resolver PR: 
3. Trait Bounds Typechecker+Monomorphizer PR: